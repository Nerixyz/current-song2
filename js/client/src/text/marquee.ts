export interface MarqueeOptions {
  speed: number;
  pauseDuration: number;
  repeatPauseDuration: number;
}

export interface MarqueeEl {
  pause: () => void;
  start: () => void;
  reset: () => void;
}

const defaultOpts: MarqueeOptions = {
  speed: 0.2,
  pauseDuration: 1200,
  repeatPauseDuration: 2000,
};

export function wrapMarquee(el: HTMLElement, opts: MarqueeOptions = defaultOpts): MarqueeEl {
  opts = { ...defaultOpts, ...opts };

  const wrap = make('div', 'mq-wrap');
  wrap.id = `mq-${el.id}`;
  const mask = make('div', 'mq-mask');
  const guard = make('div', 'mq-overflow-guard');
  const userWrap = make('div', 'mq-user-wrap');

  const parent = el.parentElement ?? document.documentElement;
  const sibling = el.nextSibling;
  el.remove();
  parent.insertBefore(wrap, sibling);

  wrap.appendChild(mask);
  mask.appendChild(guard);
  guard.appendChild(userWrap);
  userWrap.appendChild(el);

  const diffWidth = () => userWrap.clientWidth - mask.clientWidth;

  let cbID: null | number = null;
  let lastTime = performance.now();
  let pos = 0;
  let waitFrom = lastTime;
  let dir = 1;
  const applyPos = () => userWrap.style.setProperty('--marquee', `${-pos}px`);

  const actualAnimationFrame = (time: number) => {
    const deltaTime = time - lastTime;
    lastTime = time;
    const overflow = diffWidth();

    if (overflow <= 0) {
      if (pos != 0) {
        pos = 0;
        applyPos();
      }
      return;
    }

    if (waitFrom > 0) {
      if (time > waitFrom + opts.pauseDuration) {
        waitFrom = 0;
      }
    } else {
      pos += dir * ((60 * deltaTime) / 1000) * opts.speed;
      if (pos > overflow) {
        dir *= -1;
        waitFrom = time;
        pos = overflow;
      } else if (pos < 0) {
        dir *= -1;
        waitFrom = time + opts.repeatPauseDuration;
        pos = 0;
      }
    }
    applyPos();
  };
  const onAnimationFrame = (time: number) => {
    actualAnimationFrame(time);
    if (cbID) {
      cbID = requestAnimationFrame(onAnimationFrame);
    }
  };

  return {
    start: () => {
      if (!cbID) {
        lastTime = performance.now();
        cbID = requestAnimationFrame(onAnimationFrame);
      }
    },
    pause: () => {
      if (cbID) {
        cancelAnimationFrame(cbID);
        cbID = null;
      }
    },
    reset: () => {
      dir = 1;
      pos = 0;
      waitFrom = lastTime;
      applyPos();
    },
  };
}

function make(ty: keyof HTMLElementTagNameMap, ...classes: string[]): HTMLElement {
  const el = document.createElement(ty);
  el.classList.add(...classes);
  return el;
}
