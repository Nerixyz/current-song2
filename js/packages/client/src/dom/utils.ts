/**
 * Gets all elements by their id. These elements _must_ exist.
 * @param {...string} ids
 * @returns {A}
 */
export function getElements<A extends Array<HTMLElement>>(...ids: string[]): A {
  // This is unsafe but we know these elements exist when calling the function
  return ids.map(id => {
    const el = document.getElementById(id);
    if (el === null) throw new Error(`${id} does not exist`);

    return el;
  }) as A;
}
