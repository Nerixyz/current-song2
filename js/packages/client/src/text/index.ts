import { PlayInfo } from '../types';

const MAX_SAFE_STRING = 50;

interface TitleData {
  title: string;
  subtitle?: string;
}

export function extractTitleAndSub(info: PlayInfo): TitleData {
  if (info.artist) {
    return { title: info.title, subtitle: info.artist };
  }

  let title = info.title;
  if (info.source.startsWith('gsmtc') && info.source.match(/chrome|firefox|opera|brave|edge/i)) {
    console.log('xd');
    const dashIdx = title.lastIndexOf('-');
    if (dashIdx !== -1) {
      title = title.substring(0, dashIdx).trim();
    }
  }

  const dashIdx = title.indexOf('-');

  return dashIdx === -1
    ? { title }
    : {
        title: title.substring(dashIdx + 1).trim(),
        subtitle: title.substring(0, dashIdx).trim(),
      };
}

export function cleanupTitleAndSub(data: TitleData): TitleData {
  if (data.title.length > MAX_SAFE_STRING) {
    let title = cleanupTooLongString(data.title);
    let subtitle = data.subtitle;
    if (title.includes('-')) {
      const dashIdx = title.indexOf('-');
      subtitle = title.substring(0, dashIdx).trim();
      title = title.substring(dashIdx + 1).trim();
    }
    if (subtitle && subtitle?.length > MAX_SAFE_STRING) {
      subtitle = cleanupTooLongString(subtitle);
    }

    return { title, subtitle };
  } else if (data.subtitle && data.subtitle.length > MAX_SAFE_STRING) {
    return { title: data.title, subtitle: cleanupTooLongString(data.subtitle) };
  }
  return data;
}

function cleanupTooLongString(str: string): string {
  return str.replace(/(:?\(.+\)|\[.+]|{.+})/g, '').replace(/ +/, ' ').trim();
}
