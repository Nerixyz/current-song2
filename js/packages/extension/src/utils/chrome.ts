/**
 * @returns {boolean} true if the host doesn't have promise-like apis
 */
export function isChromeLike() {
  // returns 1 on Firefox and 0 on Chrome
  // => false on Firefox; true on Chrome
  return !browser.tabs.detectLanguage.length;
}
