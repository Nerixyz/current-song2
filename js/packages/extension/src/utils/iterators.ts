export function* filter<T>(iter: Iterable<T>, fn: (value: T) => boolean) {
  for (const item of iter) {
    if (fn(item)) yield item;
  }
}

export function first<T>(iter: Iterable<T>): T | null {
  const { value, done } = iter[Symbol.iterator]().next();
  return done ? null : value;
}
