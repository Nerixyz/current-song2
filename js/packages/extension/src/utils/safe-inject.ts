/**
 * Makes sure there's at most one script active
 * @param {string} key
 * @returns {boolean} true if this is the first time the script is injected.
 */
export function safeInject(key: string) {
  const idKey = Symbol.for(key);

  if ((globalThis as any)[idKey]) {
    console.debug('%c[CSO2] %cSkipping injection since the script is already injected.', 'color: red', 'color: yellow');
    return false;
  }

  Object.defineProperty(globalThis, idKey, {
    configurable: false,
    enumerable: false,
    writable: false,
    value: true,
  });

  return true;
}
