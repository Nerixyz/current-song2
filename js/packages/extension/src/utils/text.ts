export function splitTitle(title: string): { title: string; artist?: string } {
  if (title.includes('-') && !title.match(/\([^()]+-[^()]+\)/)) {
    const [first, ...second] = title.split('-');
    return { artist: first.trim(), title: second.join(' ').trim() };
  } else if (title.includes('by')) {
    // used by SoundCloud
    const [first, ...second] = title.split('by');
    return { artist: second.join(' ').trim(), title: first.trim() };
  } else {
    return { title: title.trim() };
  }
}
