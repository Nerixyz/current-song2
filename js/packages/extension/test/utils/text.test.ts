import { splitTitle } from '../../src/utils/text';

describe('splitTitle', function () {
  it('should work', function () {
    const inputs = [
      'Forsen - My Cool Video',
      'Without Dash',
      'AlienDancing by Obama',
      'LULE - AlienGathering (forsen remix)',
      'Oops - My chair broke [xd] (feat. alien)',
      'Here - Are - Multiple - Dashes',
      'Forsen - Aliens (Forsen-Remix)', // check if this is really expected
    ].map(title => splitTitle(title));

    expect(inputs).toMatchSnapshot();
  });
});
