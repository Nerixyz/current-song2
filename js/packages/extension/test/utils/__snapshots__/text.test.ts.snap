// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`splitTitle should work 1`] = `
Array [
  Object {
    "artist": "Forsen ",
    "title": " My Cool Video",
  },
  Object {
    "title": "Without Dash",
  },
  Object {
    "artist": "Obama",
    "title": "AlienDancing",
  },
  Object {
    "artist": "LULE ",
    "title": " AlienGathering (forsen remix)",
  },
  Object {
    "artist": "Oops ",
    "title": " My chair broke [xd] (feat. alien)",
  },
  Object {
    "artist": "Here ",
    "title": " Are   Multiple   Dashes",
  },
]
`;
